File: src/commands/tree.ts
/**
 * Tree command - Display sub-task dependency tree without execution
 */

import chalk from 'chalk';
import ora from 'ora';
import { readConfig } from '../lib/config.js';
import { fetchJiraIssue } from '../lib/jira.js';
import type { ParentIssue } from '../lib/linear.js';
import { fetchLinearIssue } from '../lib/linear.js';
import { readLocalSubTasksAsLinearIssues, readParentSpec } from '../lib/local-state.js';
import { renderMermaidWithTitle } from '../lib/mermaid-renderer.js';
import { resolvePaths } from '../lib/paths.js';
import { buildTaskGraph, getGraphStats } from '../lib/task-graph.js';
import { renderFullTreeOutput } from '../lib/tree-renderer.js';
import type { Backend } from '../types.js';
import { BACKEND_ID_PATTERNS } from '../types.js';

export interface TreeOptions {
  backend?: Backend;
  mermaid?: boolean;
}

/**
 * Display sub-task dependency tree for a Linear issue
 */
export async function tree(taskId: string, options: TreeOptions): Promise<void> {
  const paths = resolvePaths();
  const config = readConfig(paths.configPath);
  const backend = options.backend ?? config.backend;

  // Validate task ID format
  const pattern = BACKEND_ID_PATTERNS[backend];
  if (!pattern.test(taskId)) {
    console.error(chalk.red(`Error: Invalid task ID format for ${backend}: ${taskId}`));
    console.error(chalk.gray('Expected format: PREFIX-NUMBER (e.g., MOB-123)'));
    process.exit(1);
  }

  // Fetch parent issue with spinner (backend-aware)
  const issueSpinner = ora({
    text: backend === 'local'
      ? `Reading issue ${taskId} from local state...`
      : `Fetching issue ${taskId} from ${backend}...`,
    color: 'blue',
  }).start();

  let parentIssue: ParentIssue | null = null;

  if (backend === 'local') {
    const parentSpec = readParentSpec(taskId);
    if (parentSpec) {
      parentIssue = {
        id: parentSpec.id,
        identifier: parentSpec.identifier,
        title: parentSpec.title,
        gitBranchName: parentSpec.gitBranchName,
      };
    }
  } else if (backend === 'jira') {
    parentIssue = await fetchJiraIssue(taskId);
  } else {
    parentIssue = await fetchLinearIssue(taskId);
  }

  if (!parentIssue) {
    issueSpinner.fail(`Could not fetch issue ${taskId}`);
    process.exit(1);
  }

  issueSpinner.succeed(`${parentIssue.identifier}: ${parentIssue.title}`);
  console.log(chalk.gray(`  Branch: ${parentIssue.gitBranchName}`));

  // Read sub-tasks from local state
  const subTasksSpinner = ora({
    text: `Reading sub-tasks from local state...`,
    color: 'blue',
  }).start();

  const subTasks = readLocalSubTasksAsLinearIssues(taskId);
  if (subTasks.length === 0) {
    subTasksSpinner.warn(`No sub-tasks found for ${taskId}`);
    return;
  }

  subTasksSpinner.succeed(`Found ${subTasks.length} sub-task${subTasks.length === 1 ? '' : 's'}`);

  // Build the graph
  const graph = buildTaskGraph(parentIssue.id, parentIssue.identifier, subTasks);

  // Display ASCII tree
  console.log('');
  console.log(renderFullTreeOutput(graph));

  // Optionally display Mermaid diagram
  if (options.mermaid) {
    console.log('');
    console.log(chalk.bold('Mermaid Diagram:'));
    console.log(renderMermaidWithTitle(graph));
  }

  // Display summary stats
  const stats = getGraphStats(graph);
  console.log('');
  console.log(chalk.bold('Summary:'));
  console.log(`  Total: ${stats.total}`);
  console.log(`  Done: ${chalk.green(stats.done.toString())}`);
  console.log(`  Ready: ${chalk.blue(stats.ready.toString())}`);
  console.log(`  Blocked: ${chalk.yellow(stats.blocked.toString())}`);
  console.log(`  In Progress: ${chalk.cyan(stats.inProgress.toString())}`);
}

File: rust/mobius/src/commands/tree.rs
//! Tree command - Display sub-task dependency tree without execution

use colored::Colorize;

use crate::config::loader::read_config;
use crate::config::paths::resolve_paths;
use crate::jira::JiraClient;
use crate::local_state::{read_local_subtasks_as_linear_issues, read_parent_spec};
use crate::mermaid_renderer::render_mermaid_with_title;
use crate::tree_renderer::render_full_tree_output;
use crate::types::enums::Backend;
use crate::types::task_graph::ParentIssue;
use crate::types::task_graph::{build_task_graph, get_graph_stats};

pub fn run(task_id: &str, backend_override: Option<&str>, mermaid: bool) -> anyhow::Result<()> {
    let paths = resolve_paths();
    let config = read_config(&paths.config_path).unwrap_or_default();
    let backend: Backend = if let Some(b) = backend_override {
        b.parse().unwrap_or(config.backend)
    } else {
        config.backend
    };

    // Validate task ID format
    if !validate_task_id(task_id, &backend) {
        eprintln!(
            "{}",
            format!("Error: Invalid task ID format for {}: {}", backend, task_id).red()
        );
        eprintln!(
            "{}",
            "Expected format: PREFIX-NUMBER (e.g., MOB-123)".dimmed()
        );
        std::process::exit(1);
    }

    // Fetch parent issue
    let parent_issue: Result<ParentIssue, String> = match backend {
        Backend::Local => {
            let spec = read_parent_spec(task_id);
            spec.map(|s| ParentIssue {
                id: s.id,
                identifier: s.identifier,
                title: s.title,
                git_branch_name: s.git_branch_name,
            })
            .ok_or_else(|| format!("No local state found for {}", task_id))
        }
        Backend::Jira => {
            let rt = tokio::runtime::Runtime::new()?;
            rt.block_on(async {
                let api_err = match JiraClient::new() {
                    Ok(client) => match client.fetch_jira_issue(task_id).await {
                        Ok(issue) => return Ok(issue),
                        Err(e) => e.to_string(),
                    },
                    Err(e) => e.to_string(),
                };
                match read_parent_spec(task_id) {
                    Some(s) => Ok(ParentIssue {
                        id: s.id,
                        identifier: s.identifier,
                        title: s.title,
                        git_branch_name: s.git_branch_name,
                    }),
                    None => Err(api_err),
                }
            })
        }
        Backend::Linear => {
            let rt = tokio::runtime::Runtime::new()?;
            rt.block_on(async {
                let api_err = match crate::linear::LinearClient::new() {
                    Ok(client) => match client.fetch_linear_issue(task_id).await {
                        Ok(issue) => return Ok(issue),
                        Err(e) => e.to_string(),
                    },
                    Err(e) => e.to_string(),
                };
                match read_parent_spec(task_id) {
                    Some(s) => Ok(ParentIssue {
                        id: s.id,
                        identifier: s.identifier,
                        title: s.title,
                        git_branch_name: s.git_branch_name,
                    }),
                    None => Err(api_err),
                }
            })
        }
    };

    let parent_issue = match parent_issue {
        Ok(issue) => {
            println!("{} {}: {}", "✓".green(), issue.identifier, issue.title);
            println!(
                "  {}",
                format!("Branch: {}", issue.git_branch_name).dimmed()
            );
            issue
        }
        Err(cause) => {
            eprintln!(
                "{}",
                format!("Error: Could not fetch issue {}", task_id).red()
            );
            eprintln!("{}", format!("  Cause: {}", cause).red());
            std::process::exit(1);
        }
    };

    // Read sub-tasks from local state
    let sub_tasks = read_local_subtasks_as_linear_issues(task_id);
    if sub_tasks.is_empty() {
        println!("{}", format!("No sub-tasks found for {}", task_id).yellow());
        return Ok(());
    }

    println!(
        "{} Found {} sub-task{}",
        "✓".green(),
        sub_tasks.len(),
        if sub_tasks.len() == 1 { "" } else { "s" }
    );

    // Build the graph
    let graph = build_task_graph(&parent_issue.id, &parent_issue.identifier, &sub_tasks);

    // Display ASCII tree
    println!();
    println!("{}", render_full_tree_output(&graph));

    // Optionally display Mermaid diagram
    if mermaid {
        println!();
        println!("{}", "Mermaid Diagram:".bold());
        println!("{}", render_mermaid_with_title(&graph));
    }

    // Display summary stats
    let stats = get_graph_stats(&graph);
    println!();
    println!("{}", "Summary:".bold());
    println!("  Total: {}", stats.total);
    println!("  Done: {}", stats.done.to_string().green());
    println!("  Ready: {}", stats.ready.to_string().blue());
    println!("  Blocked: {}", stats.blocked.to_string().yellow());
    println!("  In Progress: {}", stats.in_progress.to_string().cyan());

    Ok(())
}

fn validate_task_id(task_id: &str, backend: &Backend) -> bool {
    let pattern = match backend {
        Backend::Linear => regex::Regex::new(r"^[A-Z]+-\d+$").unwrap(),
        Backend::Jira => regex::Regex::new(r"^[A-Z]+-\d+$").unwrap(),
        Backend::Local => regex::Regex::new(r"^(LOC-\d+|task-\d+)$").unwrap(),
    };
    pattern.is_match(task_id)
}
